File: Matlab Code/Classes/InstalledRegime.m
classdef InstalledRegime < handle
    %InstalledRegime Maintains data associated with a cropping regime that
    %is currently being simulated.
    %   An InstalledRegime has a list of crops (or cropNames perhaps), a list of 
    %   planting events so it knows how to get a new crop planted and a 
    %   PlantedCrop which is the crop currently planted under the regime.
    
    properties
        
        % parentSim - handle to the sim in which this regime is installed.
        parentSim
        
        % A handle to the Regime object for the regime
        regimeObject
        
        % The zone of the regime - 'primary' or 'secondary'
        % Could be made to be dependent
        zone
        
        % plantedCrops - a list of PlantedCrops, in order of when the crops
        % were planted.
        plantedCrops
        
        % outputs - outputs is a 
        % [a x m] array of Amounts containing the outputs from each month,
        % where a is the number of outputs calculated per month, and m is
        % the number of months in the simulation.
        outputs
       
        % cropIndex - the index of the current plantedCrop
        cropIndex
                
        % list of crop names used in the regime
        % Could be made to be dependent
        cropNames
        
        % initialTriggers - a cell array of trigger arrays. The cell array
        % corresponds to cropNames. Each cell contains an array of triggers
        % that match the corresponding crop's initialisation events.
        initialTriggers
        
        % installedMonth  - the monthIndex in which this regime was
        % installed.
        installedMonth
        
        % finalMonth - the monthIndex after which the regime will not be
        % installed.
        % Could be made to be dependent
        finalMonth
                        
    end
     
    % InstalledRegime constructor
    methods
        
        function obj = InstalledRegime(regimeObject, parentSim, zone)
            
            cropMgr = CropManager.getInstance;
            
            if nargin == 0
                return
            elseif nargin ~= 3
                error('Must pass 0 or 3 arguments to the InstalledRegime constructor.');
            else

                
                obj.regimeObject = regimeObject;
                obj.parentSim = parentSim;
                obj.zone = zone;
                obj.plantedCrops = PlantedCrop.empty(1, 0);
                obj.outputs = Amount.empty(1, 0); 
             %   obj.outputs.monthStart = Amount.empty(1, 0);
            %    obj.outputs.monthEnd = Amount.empty(1, 0);
                obj.cropIndex = 0;
                obj.cropNames = obj.regimeObject.cropNameList;
                obj.installedMonth = parentSim.monthIndex;
                obj.finalMonth = obj.regimeObject.finalYear * 12;

                % The only slightly difficult bit is the initialTriggers.
                %
                % InitialTriggers is a cell array whose indices correspond with
                % cropNames. Each cell has a list of initialTriggers that
                % should be used by the crop that corresponds to that cell.

                for i = 1:length(obj.cropNames)

                    % For each crop, we need to get the list of initialEvents.

                    % Need to combine the triggers from the regime and the crop...
                    % The way it works is that if an event is defined by the
                    % regime, use that event's trigger. Otherwise use the crop's
                    % trigger.

                    initialEvents = cropMgr.getCropsInitialEvents(obj.cropNames{i}); 

                    for j = 1:length(initialEvents)
                        % For each event in the crop check if the trigger is
                        % defined in the regime. If it is, use the trigger from the
                        % regime. Overwrite the triggers set just above.

                        % Try to get the trigger from regime. It will return an
                        % empty array if none exist.
                        obj.initialTriggers{i} = Trigger.empty(1, 0);
                        
                        regimeTrigger = obj.getRegimeTrigger(obj.cropNames{i}, initialEvents(j).name);

                        if ~isempty(regimeTrigger)
                           obj.initialTriggers{i}(end + 1) = regimeTrigger; 
                        else
                            % Set the crop's triggers as the default.
                            obj.initialTriggers{i} = [initialEvents.trigger];
                        end

                    end            
                end
                
            end            
        end
        
    end
    
    properties (Dependent)
       monthIndex
       monthDay
       currentPlantedCrop
       regimeLabel
       regimeOutputUnits
    end
    
    % get / set methods for dependent variables.
    methods

        % gets the current planted crop
        function cc = get.currentPlantedCrop(obj)
            if ~obj.cropIndex == 0
                cc = obj.plantedCrops(obj.cropIndex);
            else
                cc = PlantedCrop.empty(1, 0);
            end
        end
                        
        % Returns the current month index of the sim in which this
        % plantedCrop is a part.
        function mI = get.monthIndex(obj)
            mI = obj.parentSim.monthIndex;
        end
        
        % Returns the current month day of the sim in which this
        % plantedCrop is a part. 
        function mI = get.monthDay(obj)
            mI = obj.parentSim.monthDay;
        end
        
        function rL = get.regimeLabel(obj)
            rL = obj.regimeObject.regimeLabel;
        end
        
        function rOUs = get.regimeOutputUnits(obj)
            if ~isempty(obj.outputs)
                rOUs = [obj.outputs(:, 1).unit]; 
            else
                rOUs = Unit.empty(0, 1);
            end
        end
        
    end
    
    % Income, cost, profit methods
    methods
               
        % Calculates the monthly costs of the plantedCrops in this regime,
        % over the date range supplied by firstMonth and lastMonth
        % If asList is true, then cs is an array containing the monthly costs starting
        % from firstMonth. Otherwise, the total over the period is
        % returned.
        function cs = costs(obj, firstMonth, lastMonth, asList)
        
            if lastMonth < firstMonth
                cs = [];
                return;
            end
            
            if lastMonth > obj.monthIndex
               lastMonth = obj.monthIndex;
            end
            
            if firstMonth < 1 
                firstMonth = 1;
            end
            
            if asList
                cs = zeros(1, lastMonth - firstMonth + 1);
            else
                cs = 0;
            end
            
            for i = 1:length(obj.plantedCrops)
                cs = cs + obj.plantedCrops(i).costs(firstMonth, lastMonth, asList);
            end
            
        end
        
        % Calculates the monthly income of the plantedCrops in this regime,
        % over the date range supplied by firstMonth and lastMonth
        % If asList is true, then ins is an array containing the monthly income starting
        % from firstMonth. Otherwise, the total over the period is
        % returned.        
        function ins = income(obj, firstMonth, lastMonth, asList)
            if lastMonth < firstMonth
                ins = [];
                return;
            end
            
            if lastMonth > obj.monthIndex
               lastMonth = obj.monthIndex;
            end
            
            if firstMonth < 1 
                firstMonth = 1;
            end
            
            if asList
                ins = zeros(1, lastMonth - firstMonth + 1);
            else
                ins = 0;
            end
            
            for i = 1:length(obj.plantedCrops)
                ins = ins + obj.plantedCrops(i).income(firstMonth, lastMonth, asList);
            end
        
        end
        
        % Calculates the monthly profits of the plantedCrops in this regime,
        % over the date range supplied by firstMonth and lastMonth
        % If asList is true, then ps is an array containing the monthly profits starting
        % from firstMonth. Otherwise, the total over the period is
        % returned.
        function ps = profit(obj, firstMonth, lastMonth, asList)
            ps = obj.income(firstMonth, lastMonth, asList) - obj.costs(firstMonth, lastMonth, asList);
        end
        
    end
    
    % Methods for Simulation
    methods
        
        % Calculates the outputs for the installedRegime, based on the sim.
        % It then updates the installedRegime's outputs history
        % This mostly just wraps the function that's found in the Regime's RegimeDelegate.
        function calculateOutputs(obj)
            
            outputsColumn = obj.regimeObject.calculateOutputs(obj.parentSim);
            
            % The outputs returned from the current state will need to be added 
            % to the InstalledRegime's outputs property.

                if isempty(obj.outputs)
                    obj.outputs = outputsColumn';
                else
                    obj.outputs(:, obj.monthIndex - obj.installedMonth + 1) = outputsColumn';
                end                
            
        end
        
        function p = getRegimeParameter(obj, pname)
           p = obj.regimeObject.getRegimeParameter(pname); 
        end
        
        % Returns the Amount of type unit in the given month.
        function amt = getAmount(obj, unit, monthIndex)
            
            if (nargin == 2)
                monthIndex = obj.parentSim.monthIndex;
            end
            
            if (monthIndex > obj.monthIndex)                  
                amt = Amount.empty(1, 0);   
                return
            end
                 
            if (unit == Unit)
                amt = Amount(1, unit);
                return
            end

            % Grab the units column from outputs.
            outputColumnUnits = Unit.empty(1, 0);
            outputColumn = Amount.empty(1, 0);
            try
               outputColumn = obj.outputs(:, monthIndex - obj.installedMonth + 1);
               outputColumnUnits = [outputColumn.unit] ;
            catch                
            end
            
            % Work out the index of the matching column
            ix = find((outputColumnUnits == unit), 1, 'first');
            
            if isempty(ix)
               amt = Amount.empty(1, 0); 
            else
               amt = outputColumn(ix);
            end
            
        end
        
        % Like getAmount, but returns an array as long as the
        % installedRegime has existed.
        function amts = getAmounts(obj, unit)
        
            % Work out the index of the matching column
            ix = find(([obj.outputs(:, 1).unit] == unit), 1, 'first');
            
            if obj.finalMonth > obj.parentSim.monthIndex
               lastCol = obj.parentSim.monthIndex - obj.installedMonth + 1;
            else
               lastCol = obj.finalMonth - obj.installedMonth + 1; 
            end
            
            if isempty(ix)
               if (unit == Unit)                
                   amts = repmat(Amount(1, unit), 1, lastCol);                
               else
                   amts = Amount.empty(1, 0); 
               end
               return
            else
               amts = obj.outputs(ix, 1:lastCol);
            end        
        end
        
        % This function checks if there is not already a crop planted under
        % this installed regime and if not, then it tests the regime's
        % initialisation triggers in order. If one of the triggers is
        % triggered, then the crop corresponding to that trigger is
        % initialised.
        function plantIfPossible(obj)
            
            % Check if there is a crop already planted.
            
            if ~(obj.cropIndex == 0) 
                % Then we already have a crop planted.
                % Note that when a plantedCrop's destruction event is
                % triggered the SimulationManager sets the cropIndex back
                % to zero before the month ends.
                return
            end
            
            % If not, then test the initialisation triggers.
            % If one of these is triggered, then initialise that crop.
            
            % The difficult question is: How do we then initialise the
            % crop?
            
            % The elegant answer is that we simply create a PlantedCrop and
            % pass the constructor the relevant details it needs.
            % So what does it need? For a start it needs to know which
            % planting event was triggered. So perhaps we just pass the
            % plantingEventIndex. Then it will need things like the crop
            % object handle, a reference to the installedRegime, etc.
            %
            % But if the PlantedCrop comes from the CropManager, the
            % CropManager can supply all that.
            
            cropMgr = CropManager.getInstance;

            for i = 1:length(obj.initialTriggers)
                for j = 1:length(obj.initialTriggers{i})
                    
                    if obj.parentSim.isTriggered(obj.initialTriggers{i}(j))
                        
                        
                        
                        
                        % How do we get a planted crop? From the
                        % CropManager of course!
                        cropName = obj.cropNames{i};
                        initialisationEventIndex = j;
                        
                        pC = cropMgr.getPlantedCrop(cropName, obj, obj.parentSim, initialisationEventIndex);
                        
                        % Add the planted crop.
                        obj.cropIndex = length(obj.plantedCrops) + 1;
                        obj.plantedCrops(obj.cropIndex) = pC;
                        
                        % May need to fix up the units for the planting
                        % cost. Just redo them.
                        % If the costItem quantity returns non-zero as a
                        % regime output, use that value in the quantity.
                        obj.calculateOutputs;
                        ciQuantityUnit = pC.occurrences(1).costItems(1).quantity.unit;
                        amt = obj.getAmount(ciQuantityUnit);
                        if ~isempty(amt)
                            pC.occurrences(1).costItems(1).quantity.number = amt.number;
                        end
                        
                        % Return after the first trigger goes off. We don't
                        % want lots of crops being planted!
                        return
                        
                    end
                end                
            end
            
        end
        
        % Returns a trigger if it has been defined in the regime for the
        % crop and the event. Returns an empty trigger if none found.
        function rT = getRegimeTrigger(obj, cropName, eventName)
            
            rT = Trigger.empty(1, 0); 
            
            % Get the crop names
            cETs = obj.regimeObject.cropEventTriggers;
            cropNamesInRegime = {cETs.cropName};
            
            
            % Get the index of the crop in the regime
            regimeCropIndex = find(strcmp(cropNamesInRegime, cropName), 1, 'first');
            
            if isempty(regimeCropIndex)
                return
            else
                
                eTs = cETs(regimeCropIndex).eventTriggers;
                
                eventIndex = find(strcmp({eTs.eventName}, eventName), 1, 'first');
                if ~isempty(eventIndex)
                    ret = eTs(eventIndex);
                    rT = ret.trigger;
                end                
            end

        end
    end
    
end

File: imagine/regime/installed_regime.py
import numpy as np

from imagine.crop.crop_manager import CropManager
from imagine.core.amount import Amount
from dotwiz import DotWiz


class InstalledRegime:

    def __init__(self, regime_object, parent_sim, zone):
        self.parent_sim = parent_sim
        self.regime_object = regime_object
        self.zone = zone
        self.planted_crops = []
        self.outputs = parent_sim.sim_store.regime_store[self.regime_object.regime_label].outputs
        self.crop_index = -1
        self.crop_names = self.regime_object.crop_name_list
        self.installed_month = parent_sim.month_index
        self.installed_year = int(self.installed_month // 12)
        self.final_month = self.regime_object.final_year * 12 - 1
        self.initial_event_triggers = None

        self._setup_initial_triggers()

    @property
    def month_index(self):
        return self.parent_sim.month_index

    @property
    def month_day(self):
        return self.parent_sim.month_day

    @property
    def current_planted_crop(self):
        if self.planted_crops:
            last_planted_crop = self.planted_crops[-1]
            if last_planted_crop.destroyed_month is None:
                return last_planted_crop

        return None

    @property
    def regime_label(self):
        return self.regime_object.regime_label

    @property
    def regime_output_units(self):
        return [output.unit for output in self.outputs[0]] if self.outputs is not None else []

    def costs(self, first_month, last_month, as_list):
        if last_month < first_month:
            return []

        last_month = min(last_month, self.month_index)
        first_month = max(first_month, 1)

        if as_list:
            costs = [0] * (last_month - first_month + 1)
        else:
            costs = 0

        for planted_crop in self.planted_crops:
            costs += planted_crop.costs(first_month, last_month, as_list)

        return costs

    def income(self, first_month, last_month, as_list):
        if last_month < first_month:
            return []

        last_month = min(last_month, self.month_index)
        first_month = max(first_month, 1)

        if as_list:
            income_values = [0] * (last_month - first_month + 1)
        else:
            income_values = 0

        for planted_crop in self.planted_crops:
            income_values += planted_crop.income(first_month, last_month, as_list)

        return income_values

    def profit(self, first_month, last_month, as_list):
        return self.income(first_month, last_month, as_list) - self.costs(first_month, last_month, as_list)

    def calculate_outputs(self):
        outputs_column = self.regime_object.calculate_outputs(self.parent_sim)

        if self.outputs is not None:
            self.outputs[self.parent_sim.month_index, :] = outputs_column

    def get_regime_parameter(self, pname):
        return self.regime_object.get_regime_parameter(pname)

    def get_amount(self, unit, month_index=None):
        if month_index is None:
            month_index = self.parent_sim.month_index

        if month_index > self.month_index:
            return []

        if unit is None:
            return [Amount(1, None)]

        try:
            output_column = self.outputs[month_index, :]
            # output_column = [output[0] for output in self.outputs[month_index - self.installed_month]]
            output_units = [output.unit for output in output_column]
        except IndexError:
            return []

        try:
            idx = output_units.index(unit)
            return output_column[idx]
#            return [output[idx] for output in output_column]
        except ValueError:
            return []

    def get_amounts(self, unit):
        try:
            idx = [output[0].unit for output in self.outputs].index(unit)
        except ValueError:
            return []

        last_col = min(self.parent_sim.month_index - self.installed_month + 1, self.final_month - self.installed_month + 1)
        return [output[idx] for output in self.outputs[:last_col]]

    def plant_if_possible(self):
        crop_manager = CropManager.get_instance()

        for crop_name, event_triggers in self.initial_event_triggers.items():
            for event in event_triggers.events:
                trig = event_triggers.triggers[event.name]
                if trig.is_triggered(self.parent_sim, self):
                    planted_crop = crop_manager.get_planted_crop(crop_name, self, self.parent_sim, event)

                    # self.crop_index = len(self.planted_crops) + 1
                    self.planted_crops.append(planted_crop)

                    # May need to fix up the units for the planting
                    # cost. Just redo them.
                    # If the cost item quantity returns non-zero as a
                    # regime output, use that value in the quantity.
                    self.calculate_outputs()
                    for ci in planted_crop.occurrences[0].cost_items:
                        ci_quantity_unit = ci.quantity.unit
                        amt = self.get_amount(ci_quantity_unit)
                        if amt:
                            ci.quantity.number = amt.number

                    return

        # for event in eg.events:
        #     trig = eg.triggers[event.name]
        #     if trig.is_triggered(sim, self):
        #         oc = self.process_event(event, sim, fake_out)
        #         if fake_out:
        #             ocs.append(oc)
        # return ocs
        #
        #
        # if self.crop_index != 0:
        #     return
        #
        # crop_manager = CropManager.get_instance()
        #
        # for i in range(len(self.initial_triggers)):
        #     for j in range(len(self.initial_triggers[i])):
        #         for trigger in self.initial_triggers[i][j]:
        #             if self.parent_sim.is_triggered(trigger):
        #                 crop_name = self.crop_names[i]
        #                 initialisation_event_index = j
        #
        #                 planted_crop = crop_manager.get_planted_crop(crop_name, self, self.parent_sim,
        #                                                              initialisation_event_index)
        #
        #                 self.crop_index = len(self.planted_crops) + 1
        #                 self.planted_crops.append(planted_crop)
        #
        #                 # May need to fix up the units for the planting
        #                 # cost. Just redo them.
        #                 # If the cost item quantity returns non-zero as a
        #                 # regime output, use that value in the quantity.
        #                 self.calculate_outputs()
        #                 ci_quantity_unit = planted_crop.occurrences[0].cost_items[0].quantity.unit
        #                 amt = self.get_amount(ci_quantity_unit)
        #                 if amt:
        #                     planted_crop.occurrences[0].cost_items[0].quantity.number = amt[0].number
        #
        #                 return

    def _setup_initial_triggers(self):

        # Implementing event_triggers similar to PlantedCrop.
        # But it's not quite the same. Because we need to organise the initial events for all the regime crops.
        self.initial_event_triggers = DotWiz()

        # initial_event_triggers[crop_name].events is a list of the initial events for that crop.
        # initial_event_triggers[crop_name].triggers[event_name] is the trigger for that event.
        crop_manager = CropManager.get_instance()

        for crop_name in self.crop_names:
            # Each crop has a dict of events.
            # The initial events come from the crops via the crop_manager.
            # The triggers start of empty. We'll fill it in.
            self.initial_event_triggers[crop_name] = {
                'events': crop_manager.get_crops_initial_events(crop_name),
                'triggers': {}
            }

            # If the event permits regime definition, get the regime trigger and use it if not None.
            # The regime has the event configuration and can provide its trigger.
            for event in self.initial_event_triggers[crop_name].events:
                if event.status.deferred_to_regime or event.status.regime_redefinable:
                    trig = self.regime_object.get_regime_trigger(crop_name, event.name)
                    if not trig:
                        trig = event.trigger
                self.initial_event_triggers[crop_name].triggers[event.name] = trig


    # def get_regime_trigger(self, crop_name, event_name):
    #     return self.regime_object.get_regime_trigger(crop_name, event_name)
        # try:
        #
        #     trig = self.regime_object.crop_event_triggers[crop_name][event_name]
        # except KeyError:
        #     trig = None
        # return trig
        #
        # crop_event_triggers = self.regime_object.crop_event_triggers
        #
        # try:
        #     regime_crop_index = [crop_event.crop_name for crop_event in crop_event_triggers].index(crop_name)
        # except ValueError:
        #     return None
        #
        # event_triggers = crop_event_triggers[regime_crop_index].event_triggers
        #
        # try:
        #     event_index = [event_trigger.event_name for event_trigger in event_triggers].index(event_name)
        # except ValueError:
        #     return None
        #
        # return event_triggers[event_index].trigger
        #
